### PR TITLE
fix compilation on some CPUs and Windows

### PR DESCRIPTION
**Pull request by usamoi**
This patch contains two part:

* use cshim for sigsetjmp on CPUs/OSes that are not supported by `cee-scape`'s asm (needed by #1897 and Windows)
* fix linking error on Windows in #1133

### CODE REVIEW

File: pgrx-macros/src/rewriter.rs [hunk 1]
```diff
@@ -135,19 +136,52 @@ fn foreign_item_fn(func: &ForeignItemFn, abi: &syn::Abi) -> syn::Result<proc_mac
...
+                    } else {
+                        false
+                    }
+                } else {
+                    false
+                }
+            } else {
+                false
+            }
+        }
```
Comment (workingjubilee) [inline pgrx-macros/src/rewriter.rs]: HADOUKEN!

File: pgrx-macros/src/rewriter.rs [hunk 2]
```diff
@@ -135,19 +136,52 @@ fn foreign_item_fn(func: &ForeignItemFn, abi: &syn::Abi) -> syn::Result<proc_mac
     let arg_list = rename_arg_list(&func.sig)?;
     let arg_list_with_types = rename_arg_list_with_types(&func.sig)?;
     let return_type = func.sig.output.clone();
+    let link_with_cshim = func.attrs.iter().any(|attr| match &attr.meta {
+        syn::Meta::NameValue(kv) => {
+            if kv.path.get_ident().map(|x| x.to_string()) == Some("link_name".to_string()) {
```
Comment (workingjubilee) [inline pgrx-macros/src/rewriter.rs]: Ident implements PartialEq for str, you don't need to `.to_string()` this

File: pgrx-macros/src/rewriter.rs [hunk 3]
```diff
@@ -135,19 +136,52 @@ fn foreign_item_fn(func: &ForeignItemFn, abi: &syn::Abi) -> syn::Result<proc_mac
...
+                if let syn::Expr::Lit(lit) = &kv.value {
+                    if let syn::Lit::Str(value) = &lit.lit {
```
Comment (workingjubilee) [inline pgrx-macros/src/rewriter.rs]: Just do the `syn` imports here so we can make this
```suggestion
                if let Expr::Lit(ExprLit { lit: Lit::Str(value), .. }) = &kv.value {
```

File: pgrx-macros/src/rewriter.rs [hunk 4]
```diff
@@ -135,19 +136,52 @@ fn foreign_item_fn(func: &ForeignItemFn, abi: &syn::Abi) -> syn::Result<proc_mac
     let arg_list = rename_arg_list(&func.sig)?;
     let arg_list_with_types = rename_arg_list_with_types(&func.sig)?;
     let return_type = func.sig.output.clone();
+    let link_with_cshim = func.attrs.iter().any(|attr| match &attr.meta {
+        syn::Meta::NameValue(kv) => {
```
Comment (workingjubilee) [inline pgrx-macros/src/rewriter.rs]: the `if` should be able to be combined into the match arm, something like:
```suggestion
        syn::Meta::NameValue(kv) if kv.path.get_ident().filter(|x| x == "link_name").is_some() => {
```

**Review by workingjubilee — CHANGES_REQUESTED** [review]

**Comment by usamoi** [comment]
HADOUKEN has gone.

File: pgrx-bindgen/src/build.rs [hunk 5]
```diff
@@ -1116,18 +1125,11 @@ fn apply_pg_guard(items: &Vec<syn::Item>) -> eyre::Result<proc_macro2::TokenStre
...
-                    |item| match item {
-                        ForeignItem::Fn(func) => extern_funcs.push(func),
-                        item => others.push(item),
-                    },
-                );
+                let items = block.items.iter().filter(|&item| !is_blocklisted_item(item));
                 out.extend(quote! {
                     #[pgrx_macros::pg_guard]
-                    #abi { #(#extern_funcs)* }
+                    #abi { #(#items)* }
```
Comment (workingjubilee) [inline pgrx-bindgen/src/build.rs:1131]: Ah, so we're now handling `extern static`s under this. I see.

File: pgrx-macros/src/rewriter.rs [hunk 6]
```diff
@@ -135,19 +136,44 @@ fn foreign_item_fn(func: &ForeignItemFn, abi: &syn::Abi) -> syn::Result<proc_mac
...
+                false
+            }
+        }
+        _ => false,
+    });
+    let link = if link_with_cshim {
+        quote! {}
+    } else {
+        quote! { #[cfg_attr(target_os = "windows", link(name = "postgres"))] }
+    };
```
Comment (workingjubilee) [inline pgrx-macros/src/rewriter.rs:153]: Hrm. What should I be seeing and when? I am trying to force this to trigger while building the bindings and it doesn't seem to be happening?

Comment (workingjubilee) [reply pgrx-macros/src/rewriter.rs:153]: Ah, wait, this will only happen during macro expansion, sec...

Comment (workingjubilee) [reply pgrx-macros/src/rewriter.rs:153]: ...which happens after `cfg`, hmmm...

Comment (workingjubilee) [reply pgrx-macros/src/rewriter.rs:153]: Okay, forced this through `cargo expand` after some fiddling and it works as expected.

File: pgrx-pg-sys/src/submodules/ffi.rs [hunk 7]
```diff
@@ -9,7 +9,64 @@
...
 #![deny(unsafe_op_in_unsafe_fn)]
 
-use cee_scape::SigJmpBufFields;
+#[cfg(not(all(
+    any(target_os = "linux", target_os = "macos"),
+    any(target_arch = "x86_64", target_arch = "aarch64")
+)))]
+mod cee_scape {
+    #[cfg(not(feature = "cshim"))]
+    compile_error!("target platform cannot work without feature cshim");
```
Comment (workingjubilee) [inline pgrx-pg-sys/src/submodules/ffi.rs:18]: Hm. Is there a reason we're not just using the `use_c_to_interface_with_setjmp` feature of `cee_scape`?

Comment (usamoi) [reply pgrx-pg-sys/src/submodules/ffi.rs:18]: 1. To support more CPUs and Windows, we need a patch of `cee_scape`.
2. In `postgres/src/include/c.h`, sigsetjmp is defined as:
```c
/*
 * When there is no sigsetjmp, its functionality is provided by plain
 * setjmp.  We now support the case only on Windows.  However, it seems
 * that MinGW-64 has some longstanding issues in its setjmp support,
 * so on that toolchain we cheat and use gcc's builtins.
 */
#ifdef WIN32
#ifdef __MINGW64__
typedef intptr_t sigjmp_buf[5];
#define sigsetjmp(x,y) __builtin_setjmp(x)
#define siglongjmp __builtin_longjmp
#else							/* !__MINGW64__ */
#define sigjmp_buf jmp_buf
#define sigsetjmp(x,y) setjmp(x)
#define siglongjmp longjmp
#endif							/* __MINGW64__ */
#endif							/* WIN32 */
```
I feel it's better to include the header and use `sigsetjmp` in `pgrx-cshim.c` than adding a `__builtin_setjmp` version of `call_closure_with_setjmp` in `cee-scape` and gating the variants by cfg...

Comment (workingjubilee) [reply pgrx-pg-sys/src/submodules/ffi.rs:18]: Ah, interesting.

**Review by workingjubilee — APPROVED** [review]
Thank you!

**Pull request merged**